spatialpy/ssa_sdpd-c-simulation-engine/src/simulate_rdme.cpp

```diff
@@ -0,0 +1,876 @@
+            fprintf(fp, "%zu",irD[i]);
+        }
+        fprintf(fp, "]\n");
+        fprintf(fp, "jcD = [");
+        for(i=0;i<jcD_ndx;i++){
+            if(i>0){ fprintf(fp,",");}
+            fprintf(fp, "%zu",jcD[i]);
+        }
+        fprintf(fp, "]\n");
+        fprintf(fp, "prD = [");
+        for(i=0;i<prD_ndx;i++){
+            if(i>0){ fprintf(fp,",");}
+            fprintf(fp, "%e",prD[i]);
+        }
+        fprintf(fp, "]\n");
+        fprintf(fp, "D = scipy.sparse.csc_matrix(prD,irD,jcD)\n");
+        fclose(fp);
+
+    }**/
+
+    /**************************************************************************/
+    /**
+    void nsm_core__destroy_diffusion_matrix(rdme_t*rdme){
+        //free(rdme->irD);
+        //free(rdme->jcD);
+        //free(rdme->prD);
+    }
+    **/
+    /**************************************************************************/
+    // Update to use priority queue
+    void nsm_core__take_step(ParticleSystem*system, double current_time, double step_size){
+        // rdme_t*rdme = system->rdme;
+        double tt = current_time;
+        double end_time = current_time + step_size;
+        double totrate,cum,rdelta,rrdelta;
+        int event,errcode = 0;
+        long unsigned int re, spec = 0 ;
+        Particle*subvol;
+        size_t i,j = 0;
+        //double old_rrate = 0.0,old_drate = 0.0;
+        double rand1,rand2,cum2,old;
+        double vol,diff_const;
+        Particle *dest_subvol = NULL;
+        NeighborNode *nn = NULL;
+
+
+        // Check the integrety of the heap
+    //    ordered_node_t*on,*bon,*aon;
+    //    printf("========================================================\n");
+    //    on= system->rdme->heap->head;
+    //    bon=NULL;
+    //    aon=system->rdme->heap->head->next;
+    //    while(on!=NULL){
+    //        if( on->prev != bon ||
+    //            on->next != aon ){
+    //            printf("next/prev mismatch\n");exit(1);
+    //        }
+    //        on = on->next;
+    //        if(bon==NULL){bon=system->rdme->heap->head;}else{bon=bon->next;}
+    //        if(aon!=NULL){aon=aon->next;}
+    //    }
+    //    if(system->rdme->heap->tail != bon){
+    //        printf("tail mismatch\n");exit(1);
+    //    }
+    //    printf("========================================================\n");
+
+
+        std::pair<double,int> timeRxnPair;
+        int reactionIndex;
+        /* Main loop. */
+        while(tt <= end_time){
+
+            /* Get the subvolume in which the next event occurred.
+             This subvolume is on top of the heap. */
+            //told = tt;
+            //tt   = rdme->rtimes[0];
+            //subvol = rdme->node[0];
+            //subvol = system->event_v.front().data;
+            //tt = system->event_v.front().tt;
+
+            timeRxnPair = system->rdme_event_q.selectReaction();
+            tt = timeRxnPair.first;
+            subvol_index = timeRxnPair.second;
+            subvol = system->particles[subvol_index];
```

Comment:
`no suitable conversion function from "Spatialpy::Particle" to "Spatialpy::Particle *" exist`